crates/ra_analysis/src/imp.rs

```diff
@@ -258,6 +258,17 @@ impl AnalysisImpl {
 
         let mut ret = vec![];
 
+        if let Some(binding) = find_node_at_offset::<ast::BindPat>(syntax, offset) {
+            let decl = DeclarationDescriptor::new(binding);
+
+            ret.push((file_id, decl.range));
+
+            ret.extend(decl.find_all_refs().into_iter()
+                .map(|ref_desc| (file_id, ref_desc.range )));
+
+            return ret;
+        }
+
         // Find the symbol we are looking for
         if let Some(name_ref) = find_node_at_offset::<ast::NameRef>(syntax, offset) {
```

Comment:
At this point I would like to use the `DeclarationDescriptor` machinery and was thinking of a way to work backwards to an `ast::BindingPat`.  Something like:

```rust
 let maybe_binding = find_node_at_offset::<ast::BindPat>(syntax, offset)
            .or_else(||
                find_node_at_offset::<ast::NameRef>(syntax, offset)
                    .and_then(|name_ref| resolve_local_name(name_ref))
                    .and_then(|resolved| {
                        eprintln!("** resolved: {:?}", resolved);
                        let n = find_node_at_offset::<ast::BindPat>(syntax, resolved.1.start());
                        eprintln!("**** n: {:?}", n);
                        n
                    })
            );

        if maybe_binding.is_none() {
            eprintln!("** maybe_binding is none");
            return ret;
        }

        let binding = maybe_binding.unwrap();

        let decl = DeclarationDescriptor::new(binding);
```

Which would simplify this function quite a bit.  However `find_node_at_offset::<ast::BindPat>(syntax, resolved.1.start());` always returns `None`. I'm sure that I am missing something obvious. Dumping the syntax for my test shows that I should be resolving an `ast::BindPat` there.

Reply:
Okay actually... if I change the offset to: `resolved.1.end() it seems to work()`